clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -968,13 +1035,23 @@ private void transitionTo(State target) {
     }
 
     private void transitionTo(State target, RuntimeException error) {
+        transitionTo(target, error, InvalidStateTransitionHandler.SET_FATAL_STATE);
+    }
+
+    private void transitionTo(State target, RuntimeException error, InvalidStateTransitionHandler invalidStateTransitionHandler) {
         if (!currentState.isTransitionValid(currentState, target)) {
             String idString = transactionalId == null ?  "" : "TransactionalId " + transactionalId + ": ";
-            throw new IllegalStateException(idString + "Invalid transition attempted from state "
-                    + currentState.name() + " to state " + target.name());
-        }
+            String message = idString + "Invalid transition attempted from state "
+                    + currentState.name() + " to state " + target.name();
 
-        if (target == State.FATAL_ERROR || target == State.ABORTABLE_ERROR) {
+            // See InvalidStateTransitionHandler above for more detail.
+            if (invalidStateTransitionHandler == InvalidStateTransitionHandler.THROW_EXCEPTION) {
+                throw new IllegalStateException(message);
+            } else {
+                lastError = new IllegalStateException(message);
+                target = State.FATAL_ERROR;
```

Comment:
Would it make sense to also throw after we have transitioned to FATAL_ERROR?

Reply:
When I change the code to _always_ set `currentState` to `FATAL_STATE`, the `TransactionManagerTest.testRetryInitTransactionsAfterTimeout` test fails. From the look of that test (and my understanding), we don't want to "poison" the transaction manager in all cases.

Reply:
Jason is saying something else though, right? Isn't he saying that we should always throw the error too (after transitioning to fatal error)?

Reply:
Ah, OK.

Is this the desired logic:

1. For `Producer` API calls, _don't_ update the state to `FATAL_STATE`, but do throw an exception
2. For background threads, update the status to `FATAL_STATE` *and* throw an exception

Does that sound correct @jolshan @hachikuji?

Reply:
I think this is what Jason meant, but would be good to confirm. 😅